src/storage-client/src/controller.rs

```diff
@@ -2505,6 +2507,72 @@ where
         self.reconcile_managed_collection(id, updates).await;
     }
 
+    /// Effectively truncates the source status history shard except for the most recent update from
+    /// each ID.
+    async fn initialize_source_status_history(&mut self) {
+        let id = self.state.introspection_ids[&IntrospectionType::SourceStatusHistory];
+
+        let (occurred_at, _) = healthcheck::MZ_SOURCE_STATUS_HISTORY_DESC
+            .get_by_name(&ColumnName::from("occurred_at"))
+            .expect("schema has not changed");
+
+        let (source_id, _) = healthcheck::MZ_SOURCE_STATUS_HISTORY_DESC
+            .get_by_name(&ColumnName::from("source_id"))
+            .expect("schema has not changed");
+
+        let mut rows = vec![];
+
+        match self.state.collections[&id]
+            .write_frontier
+            .elements()
+            .iter()
+            .min()
+        {
+            Some(f) if f > &T::minimum() => {
+                let as_of = f.step_back().unwrap();
+
+                rows.extend(self.snapshot(id, as_of).await.unwrap());
+            }
+            // If collection is closed or the frontier is the minimum, we cannot
+            // or don't need to truncate (respectively).
+            _ => {}
+        }
+
+        let mut most_recent_status = BTreeMap::new();
+
+        for (row, diff) in rows.iter() {
+            mz_ore::soft_assert!(
+                *diff == 1,
+                "only know how to operate over consolidated data"
+            );
+
+            let d = row.unpack();
+            let source_id = d[source_id];
+            let occurred_at = d[occurred_at];
+
+            let (most_recent, unpacked_row) = most_recent_status
+                .entry(source_id)
+                .or_insert((occurred_at.clone(), d.clone()));
+
+            if occurred_at > *most_recent {
+                *most_recent = occurred_at;
+                *unpacked_row = d;
+            }
+        }
+
+        let mut row_buf = Row::default();
+        let updates = most_recent_status
+            .into_iter()
+            .map(|(_, (_, unpacked_row))| {
+                let mut packer = row_buf.packer();
+                packer.extend(unpacked_row.into_iter());
+                (row_buf.clone(), 1)
+            })
+            .collect();
+
+        self.reconcile_managed_collection(id, updates).await;
```

Comment:
We're doing duplicate work because `reconcile_managed_collection` will again take a snapshot of the shard. We already have the content in `rows`, so could just filter away those that we wan't to retain and apply retractions using `append_to_managed_collection`.

You probably are already aware and just wanted a review of the general approach, though. 🕊️

Reply:
Yes, but ty for the pointer!